src/core/dense_set.h

```diff
@@ -0,0 +1,473 @@
+// for 75% utilization (1/0.75 buckets): N*1.33*8 + N*24 = 35N
+//
+// This class uses 8 bytes per bucket (similarly to dictEntry*) but it used it for both
+// links and keys. For most cases, we remove the need for another redirection layer
+// and just store the key, so no "dictEntry" allocations occur.
+// For those cells that require chaining, the bucket is
+// changed in run-time to represent a linked chain.
+// Additional feature - in order to to reduce collisions, we insert items into
+// neighbour cells but only if they are empty (not chains). This way we reduce the number of
+// empty (unused) spaces at full utilization from 36% to ~21%.
+// 100% utilized table requires: N*8 + 0.2N*16 = 11.2N bytes or ~20 bytes savings.
+// 75% utilization: N*1.33*8 + 0.12N*16 = 13N or ~22 bytes savings per record.
+// with potential replacements of hset/zset data structures.
+// static_assert(sizeof(dictEntry) == 24);
+
+class DenseSet {
+ public:
+  bool Reserve(size_t sz);
+
+ protected:
+  // Virtual functions to be implemented for generic data
+  virtual uint64_t Hash(const void*) const = 0;
+  virtual bool Equal(const void*, const void*) const = 0;
+  virtual size_t ObjectAllocSize(const void*) const = 0;
+  virtual void Clear();
+
+  bool Add(void*);
+  bool Contains(const void*) const;
+  void* Erase(void*);
+  void* Pop();
+
+ private:
+  struct DenseLinkKey;
+  // we can assume that high 12 bits of user address space
+  // can be used for tagging. At most 52 bits of address are reserved for
+  // some configurations, and usually it's 48 bits.
+  // https://www.kernel.org/doc/html/latest/arm64/memory.html
+  static constexpr size_t kLinkBit = 1ULL << 52;
+  static constexpr size_t kDisplaceBit = 1ULL << 53;
+  static constexpr size_t kDisplaceDirectionBit = 1ULL << 54;
+  static constexpr size_t kTagMask = 4095ULL << 51;  // we reserve 12 high bits.
+
+  struct DensePtr {
+    explicit DensePtr(void* p = nullptr) : ptr_(p) {
+    }
+
+    bool IsObject() const {
+      return (uintptr_t(ptr_) & kLinkBit) == 0;
+    }
+
+    bool IsLink() const {
+      return (uintptr_t(ptr_) & kLinkBit) == kLinkBit;
+    }
+
+    bool IsEmpty() const {
+      return ptr_ == nullptr;
+    }
+
+    void* Raw() const {
+      return (void*)(uintptr_t(ptr_) & ~kTagMask);
+    }
+
+    bool IsDisplaced() const {
+      return (uintptr_t(ptr_) & kDisplaceBit) == kDisplaceBit;
+    }
+
+    void SetLink(DenseLinkKey* lk) {
+      ptr_ = (void*)(uintptr_t(lk) | kLinkBit);
+    }
+
+    void SetDisplaced(int direction) {
+      DCHECK(direction == -1 || direction == 1);
+      ptr_ = (void*)(uintptr_t(ptr_) | kDisplaceBit);
+      if (direction == 1) {
+        ptr_ = (void*)(uintptr_t(ptr_) | kDisplaceDirectionBit);
+      }
+    }
+
+    void ClearDisplaced() {
+      ptr_ = (void*)(uintptr_t(ptr_) & ~kDisplaceBit);
+      ptr_ = (void*)(uintptr_t(ptr_) & ~kDisplaceDirectionBit);
+    }
+
+    int GetDisplacedDirection() const {
```

Comment:
nice! 
please add a comment saying that -1 means that the item is located to the left of its home bucket and 1 is to the right.